addons/binding/org.openhab.binding.snapcast/src/main/java/org/openhab/binding/snapcast/internal/protocol/SnapcastController.java

```diff
@@ -0,0 +1,421 @@
+     * @param responseHandler Handler for processing the response
+     */
+    public <T> void sendRequest(String method, @Nullable Identifiable params, Class<T> responseType,
+            Consumer<T> responseHandler) {
+        final HandlerRef<T> handlerRef = new HandlerRef<>(responseType, responseHandler, params);
+
+        int id;
+        synchronized (responseHandlers) {
+            do {
+                id = requestId.nextInt(Integer.MAX_VALUE);
+            } while (responseHandlers.containsKey(id));
+            responseHandlers.put(id, handlerRef);
+        }
+
+        OutputMessage<Object> msg = new OutputMessage<>();
+        msg.id = id;
+        msg.jsonrpc = "2.0";
+        msg.method = method;
+        msg.params = params;
+
+        SocketHandler sh = this.socketHandler;
+        if (sh != null) {
+            sh.write(gson.toJson(msg));
+        }
+    }
+
+    /**
+     * Snapcast connection handler
+     *
+     * @author Steffen Brandemann - Initial contribution
+     */
+    private class SocketHandler extends Thread {
+
+        private final String host;
+        private final Integer port;
+
+        private @Nullable Socket socket;
+
+        public SocketHandler(String host, Integer port) {
+            this.host = host;
+            this.port = port;
+        }
+
+        @Override
+        public void run() {
+            boolean running = true;
+            while (running && !isInterrupted()) {
+                try {
+                    final Socket socket;
+                    synchronized (SocketHandler.this) {
+                        socket = this.socket;
+                    }
+                    if (socket != null && socket.isConnected() && !socket.isClosed()) {
+                        try {
+                            String msg;
+                            BufferedReader reader = new BufferedReader(new InputStreamReader(socket.getInputStream()));
+                            while ((msg = reader.readLine()) != null) {
+                                parseInputMessage(msg);
+                            }
+                        } catch (SocketException e) {
+                            final String message = e.getMessage();
+                            if (message == null || !message.equals("Socket closed")) {
+                                throw e;
+                            }
+                        } finally {
+                            serverController.disconnected();
+                            closeSocket();
+                        }
+                    } else {
+                        try {
+                            openSocket();
+                        } catch (IOException e) {
+                            Thread.sleep(5000);
+                        }
+                    }
+                } catch (InterruptedException e) {
+                    running = false;
+                } catch (IOException e) {
+                    logger.error("snapcast connection error", e);
```

Comment:
In [E.4 of the coding guidelines](http://docs.openhab.org/developers/development/guidelines.html#e-logging) we listed what kind of logging level should be used when. Error should only be used:

> `error` logging should only be used to inform the user that something is tremendously wrong in his setup, the system cannot function normally anymore, and there is a need for immediate action. 

Please change this to use the correct logging levels and if possible change the ONLINE/OFFLINE status of thing / bridge.